[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-brightgreen.svg)](https://github.com/ag-computational-bio/bakrep/blob/main/LICENSE)

:microbe: BakRep: A searchable web repository for bacterial genomes and standardized characterizations
=======

A workflow for taxonomic classification, subtyping and annotation of the assemblies generated by _Blackwell et al., 2021_.

# :mag: Background 

The large amount of bacterial genomic data in public genome databases is an important resource for research in various fields. However, most of this data has been processed differently, making accurate comparisons challenging. _Blackwell et al._, used a uniform approach to assemble and characterize 661,405 bacterial genomes retrieved from the European Nucleotide Archive  [(ENA)](https://www.ebi.ac.uk/ena/browser/home) in November of 2018. We builded up on this resource and further analyzed the assembled genomes in a standardized manner. We conducted a robust taxonomic classification using the Genome Taxonomy Database (GTDB) and furthermore subtyped all eligible genomes via multilocus-sequence typing. In addition we annotated all genomes assigning functional categories, e.g. COG and E.C., and database cross references to public databases. The overarching goal is to make this standardised resource available to non-bioinformaticians via an interactive [website](https://bakrep.computational.bio/). This website will provide researchers with a flexible search engine to query the repository. 

# :hammer_and_wrench:  Installation  

The workflow is written in [Nextflow](https://www.nextflow.io/docs/latest/index.html) and all tools are installed via [Conda](https://conda.io/projects/conda/en/latest/user-guide/install/index.html). Python3 is also required to create the result json files.

Clone the repository:

```bash
git clone https://github.com/ag-computational-bio/bakrep
```

## Running

Run the nextflow-script:

```bash
nextflow run /nextflow/661k.nf 
--samples <metadata> 
--setupdir <setupdir>
-with-conda
```

# ðŸ—ƒï¸ Parameters 

All optional parameters:

```
--samples             Metadata file which contains sample-id, genus, species and strain. 
--setupdir            Path to folder which contains all data.
--db                  Path to database folder which contains GTDBtk, CheckM2 and Bakta database (default: $setupdir/database)
--gtdb                Path to GTDBtk database (no need, when set --db)
--checkm2db           Path to CheckM2 database (no need, when set --db)
--baktadb             Path to Bakta database (no need, when set --db)
--data                Path to input assembly files (default: $setuptdir/assemblies)
--results             Path to output folder (default: $setupdir/results)
```

# :page_facing_up: Citation

>Blackwell GA, Hunt M, Malone KM, Lima L, Horesh G, Alako BTF, Thomson NR, Iqbal Z. Exploring bacterial diversity via a curated and searchable snapshot of >archived DNA sequences. PLoS Biol. 2021 Nov 9;19(11):e3001421. doi: 10.1371/journal.pbio.3001421. PMID: 34752446; PMCID: PMC8577725.

## Tools

- [GTDBtk](https://github.com/Ecogenomics/GTDBTk)
- [Checkm2](https://github.com/chklovski/CheckM2)
- [assembly-scan](https://github.com/rpetit3/assembly-scan)
- [mlst](https://github.com/tseemann/mlst)
- [Bakta](https://github.com/oschwengers/bakta)
